rsf.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (C) 2023 by Posit Software, PBC
+package rsf
+
+import (
+	"bufio"
+	"io"
+)
+
+type Writer interface {
+	// WriteObject uses reflection and `rsf` struct tag annotations to write an object.
+	WriteObject(v any) (int, error)
+
+	// WriteSizeField writes a 4-byte field that indicates a size (usually the
+	// size in bytes of an object or value, or an array length).
+	WriteSizeField(pos int, val int, r io.Writer) (int, error)
+
+	// WriteFixedStringField writes a string of a fixed length. An error is returned
+	// if the string size does not match the provided `sz` parameter.
+	WriteFixedStringField(pos, sz int, val string, r io.Writer) (int, error)
+
+	// WriteStringField writes a variable length string. The string value will be
+	// prepended with a 4-byte size field that indicates the string length.
+	WriteStringField(pos int, val string, r io.Writer) (int, error)
+
+	// WriteBoolField writes a 1-byte (0 or 1) boolean value.
+	WriteBoolField(pos int, val bool, r io.Writer) (int, error)
+}
+
+// Reader - The Reader interface provides Read* methods analagous to the Write*
+// methods in the Writer interface. No `ReadObject` method is provided since
+// reading is likely to be customized per use case.
+type Reader interface {
+	ReadSizeField(r io.Reader) (int, error)
+	ReadFixedStringField(sz int, r io.Reader) (string, error)
+	ReadStringField(r io.Reader) (string, error)
+	ReadBoolField(r io.Reader) (bool, error)
+
```

Comment:
Good question! At this point, I think we should keep it out for two reasons:

* Support for `Peek` is dependent upon the type of reader. If you have a reader that supports `Peek`ing, you can peek anyway without adding a method to this interface.
* `Peek` doesn't affect the reader position. It seemed ideal to add `Seek` to this interface since that affects the `Pos()` value, but I don't think there's a corresponding benefit for `Peek`.